File: backend/spring-boot/src/main/java/org/bootstrapbugz/backend/user/repository/UserRepository.java
package org.bootstrapbugz.backend.user.repository;

import java.util.List;
import java.util.Optional;
import org.bootstrapbugz.backend.user.model.User;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

public interface UserRepository extends JpaRepository<User, Long> {
  @Query("select u.id from User u order by u.id")
  List<Long> findAllUserIds(Pageable pageable);

  @EntityGraph(attributePaths = "roles")
  List<User> findAllByIdIn(List<Long> ids);

  @EntityGraph(attributePaths = "roles")
  Optional<User> findWithRolesById(Long id);

  Optional<User> findByUsername(String username);

  @EntityGraph(attributePaths = "roles")
  Optional<User> findWithRolesByUsername(String username);

  Optional<User> findByEmail(String email);

  Optional<User> findByUsernameOrEmail(String username, String email);

  boolean existsByUsername(String username);

  boolean existsByEmail(String email);
}

File: backend/spring-boot-kotlin/src/main/kotlin/org/bootstrapbugz/api/user/repository/UserRepository.kt
package org.bootstrapbugz.api.user.repository

import org.bootstrapbugz.api.user.model.User
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import java.util.Optional

interface UserRepository : JpaRepository<User, Long> {
    @EntityGraph(attributePaths = ["roles"])
    override fun findAll(pageable: Pageable): Page<User>
    @EntityGraph(attributePaths = ["roles"])
    fun findWithRolesById(id: Long): Optional<User>
    fun findByUsername(username: String): Optional<User>
    @EntityGraph(attributePaths = ["roles"])
    fun findWithRolesByUsername(username: String): Optional<User>
    fun findByEmail(email: String): Optional<User>
    fun findByUsernameOrEmail(username: String, email: String): Optional<User>
    fun existsByUsername(username: String): Boolean
    fun existsByEmail(email: String): Boolean
}
